packages/core/src/index.ts

```diff
@@ -56,6 +62,11 @@ export const stepUntilConvergence = (state: State, numSteps = 10000): State => {
   while (!stateConverged(currentState)) {
```

Comment:
The color assignment for shapes depends on the spatial proximity of different shapes in the final optimized state (and the color assignment for text depends on the color assignment for shapes), so i think the color functions have to be called after the state converges (?)

I'm not sure why, but calling the color functions in this fashion instead:
```
export const interactiveDiagram = async (
  domainProg: string,
  subProg: string,
  styProg: string,
  node: HTMLElement
): Promise<void> => {
  const updateData = (state: State) => {
    const stepped = colorUninitText(colorUninitShapes(stepUntilConvergence(state)));
    node.replaceChild(
      RenderInteractive(stepped, updateData),
      node.firstChild as Node
    );
  };
  const res = compileTrio(domainProg, subProg, styProg);
  if (res.isOk()) {
    const state: State = await prepareState(res.value);
    const optimized = colorUninitText(colorUninitShapes(stepUntilConvergence(state)));
    node.appendChild(RenderInteractive(optimized, updateData));
  } else {
    throw Error(
      `Error when generating Penrose diagram: ${showError(res.error)}`
    );
  }
};
```
doesn't seem to record the newly assigned colors in the state (and the browser ui lags when I try to step to convergence).

tl;dr I'll leave the color functions in `stepToConvergence` for now, but I'll play around to see if I can integrate them more cleanly into the overall system @hypotext

Reply:
> The color assignment for shapes depends on the spatial proximity of different shapes in the final optimized state (and the color assignment for text depends on the color assignment for shapes)

Oh right, I didn't realize this. Hmm, putting the function at this location seems like it might clash with the results if the user tries to optimize colors in style - the colors will be left unassigned and optimized, and then the colors will get set again in your pass, resulting in unexpected results. 

But then, your function depends on the optimized results.

Maybe your functionality can be optionally turned on via a flag in Style? e.g. 

`settings {
    ASSIGN_COLORS = true
}
`

would trigger it?

Reply:
> Oh right, I didn't realize this. Hmm, putting the function at this location seems like it might clash with the results if the user tries to optimize colors in style - the colors will be left unassigned and optimized, and then the colors will get set again in your pass, resulting in unexpected results.

If someone tries to optimize `A.shape.color`, for instance, will `A.shape.color` still end up in `uninitializedPaths`? If not, there's nothing to worry about. I'm actually not sure about this.

> `settings { ASSIGN_COLORS = true } `

I'd be careful about adding magic values in Style.

Reply:
> If someone tries to optimize A.shape.color, for instance, will A.shape.color still end up in uninitializedPaths? If not, there's nothing to worry about. I'm actually not sure about this.

Yes, because usually colors are optimized by leaving them uninitialized in Style. Since a color is not a `Float` type, it will end up in `uninitializedPaths`. The only way to really know if something is optimized is to see if it or its constituent parts appear somewhere in the energy graph, which is pretty complicated.

> I'd be careful about adding magic values in Style.

Rather than a magic value, I would think of it as a user setting, since some users might not want it. So maybe it can be a tickbox in the inspector instead? Anyway, open to other proposed solutions here.

Reply:
After thinking more about this @wodeni, I think a Style writer actually cannot optimize a color that's left uninitialized*. That means that if the color's path appears in `uninitializedPaths`, it can be set as @heleaf is doing.

*this is because if a shape's color is not set, then by default, we **don't** make each of the color components into varying floats; they can only be constants in the optimization. If someone wants to optimize a color, they have to manually make a varying path for each component, e.g. `A.r = ?; A.g = ?; A.b = ?; A.shape = Circle { color : rgba(A.r, A.g, A.b, 1.) }`. So in a sense the invariant is due to a system decision. Whereas if we **did** make uninitialized color components into varying floats by default, then we could not rely on uninitialized colors to be not optimized.

Reply:
I think the actual edge case here is: if a color component participate in a constraint/objective (e.g. `ensure equal(A.shape.color[0], 1.0)`), then (1) this component (`A.shape.color[0]`) will appear in `varyingPaths` __and__ (2) the color itself is still in `uninitializedPaths`. 

If (2) is false, I think the current approach will work just fine. Otherwise, can we change the Style compiler such that it filters out values like `A.shape.color` from `uninitializedPaths`?